### PR TITLE
gcp slice data functions

### PR DESCRIPTION
**Pull request by colman-hartinger**
Functions to convert gcp slice data to something readable in wim. 

Notes: 

1. I used some functions from es 2019. The only browser that can't handle them is internet explorer which is defunct
2. I left out most of the static type checking since I'm not sure how we want to integrate the GCP interfaces.ts into thorjs
3. I took a functional approach to building the slice data since I'm not quite sure how it might need to be grouped in the future. If you think it be better to wrap the functions in a class let me know

### CODE REVIEW

File: src/chop/slicer/slicer.ts [hunk 1]
```diff
@@ -6,6 +6,29 @@ export enum SlicerType {
   cura = 'cura',
 }
 
+export enum BoundaryCountourType {
```
Comment (bradymadams) [inline src/chop/slicer/slicer.ts:9]: Spelling

File: src/chop/slicer/sliceData.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+function getSmartSliceLayer(layer: any, index) { 
+  const boundaryLoops = layer.boundaries;
+  const landBoundary = boundaryLoops[0].landBoundaryLoop;
+  const airBoundaries = boundaryLoops[0].airBoundaryLoops;
+  const internalContours = layer.internalContours;
+
+
+
```
Comment (bradymadams) [inline src/chop/slicer/sliceData.ts:41]: Excessive whitespace

File: src/chop/slicer/sliceData.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  const landPolygons = {
+    id:1,
```
Comment (bradymadams) [inline src/chop/slicer/sliceData.ts:43]: Whitespace

File: src/chop/slicer/sliceData.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import {BoundaryCountourType} from './slicer'
+
+
+
```
Comment (bradymadams) [inline src/chop/slicer/sliceData.ts:4]: Excessive whitespace

**Review by bradymadams — CHANGES_REQUESTED** [review]
1. Lots of missing semi-colons
2. Can you describe how this code is going to be used? Looks like it's in an intermediate state - in which case we should probably merge to an intermediate branch while you work on the rest. Should probably also target 21.1 for this work.
3. Looks like there's some missing types on some of the arguments in the new functions. I assume it would introduce a circular dependency to add the proper types. In which case - I wonder if the translation code should live in the plug-in, not thorjs.
4. I'd like to see some example screenshots of a part and the corresponding rendered slices. We should validate that it's working. If we decide to move any code to the plugin side you can hold off here until that's done.

**Comment by colman-hartinger** [comment]
>     2. Can you describe how this code is going to be used? Looks like it's in an intermediate state - in which case we should probably merge to an intermediate branch while you work on the rest. Should probably also target 21.1 for this work.
> 
>     3. Looks like there's some missing types on some of the arguments in the new functions. I assume it would introduce a circular dependency to add the proper types. In which case - I wonder if the translation code should live in the plug-in, not thorjs.
> 
>     4. I'd like to see some example screenshots of a part and the corresponding rendered slices. We should validate that it's working. If we decide to move any code to the plugin side you can hold off here until that's done.

2.  Ya I would say this in more of an intermediate state, If you cycle over all the layers calling `getSmartSliceLayer` you'll end up with the structure the slice tool expects. I still wasn't sure what the packet sent from the api was going to exactly look like and if there was a different format that wim expected so I left a bit of the functionality unimplemented.
image

3. Ya I think this code would be better served in ss-gcp-ui since a lot of the types would need to be copied from there if we wanted to keep it in thorjs. I don't think there would be any circular dependencies though just some code duplication.
4. 
image
First layer
image
Middle layer
image
Top layer

I also sent you the chop.json file for this model on slack if you want to check out some of intermediate layers but I think everything is looking as expected with all the "Exterior" "Walls", "Infill" and "Skins" in the expected locations.

**Comment by bradymadams** [comment]
> I don't think there would be any circular dependencies though just some code duplication.

The circ dependency comment was made assuming we'd try to avoid code duplication. I think this means there's a little bit of a smell here. I recommend we move the code back to the UI. We can update thorjs to maintain a concept of the slicer layer data in our format, if we want (that would probably be ideal), but the translation needs to happen where all types are available and no code duplication is necessary (UI).

The screenshots look good. I'd like to see some corresponding GCP screenshots with settings, and make sure you've dug in and compared sliced layers to the toolpaths in GCP to make sure everything looks correct. We need to validate that the "pretty pictures" are correct.

**Pull request closed without merging**